### PR TITLE
Enable others to see who the current master is

### PR DESCRIPTION
**Pull request by AlCutter**
This PR exposes the instance ID of the current master (if any) for a given election.
master

### CODE REVIEW

File: server/log_operation_manager_test.go [hunk 1]
```diff
@@ -355,6 +355,10 @@ func (te *testElection) Close(ctx context.Context) error {
 	return te.closeErr
 }
 
+func (te *testElection) GetCurrentMaster(ctx context.Context) (string, error) {
+	return "It's you!", nil
```
Comment (daviddrysdale) [inline server/log_operation_manager_test.go]: "It's a me, Mario!"

File: util/election.go [hunk 2]
```diff
@@ -29,6 +29,8 @@ type MasterElection interface {
 	ResignAndRestart(context.Context) error
 	// Close permanently stops the mastership election process.
 	Close(context.Context) error
+	// GetCurrentMaster returns the instance ID of the current elected master, if any.
```
Comment (daviddrysdale) [inline util/election.go]: Maybe add a little bit of text about "instance ID" as it's new to this interface -- is it unique and stable and therefore safe to use for election-related behaviour, or is it more of an indication that's a debugging help but not to be relied on for serious distributed system design?

Comment (daviddrysdale) [reply util/election.go]: Also, maybe have a specific documented error value that the interface can use to indicate no-leader (and map `concurrency.ErrElectionNoLeader` to it in the etcd code)?

File: util/etcd/election_test.go [hunk 3]
```diff
@@ -54,3 +60,89 @@ func TestMasterElectionThroughCommonClient(t *testing.T) {
...
+
+	// add a couple of passive observers
+	ob1 := mustCreateClientFor(t, e.Config().LCUrls[0].String())
+	defer ob1.Close()
+	ob2 := mustCreateClientFor(t, e.Config().LCUrls[0].String())
+	defer ob2.Close()
+
+	ctx := context.Background()
+	facts := make([]*ElectionFactory, 0)
+	facts = append(facts, NewElectionFactory("serv1", client, "trees/"))
```
Comment (daviddrysdale) [inline util/etcd/election_test.go]: nit: use a constant for `"trees/"`

File: util/etcd/election_test.go [hunk 4]
```diff
@@ -54,3 +60,89 @@ func TestMasterElectionThroughCommonClient(t *testing.T) {
...
+	defer client2.Close()
+
+	// add a couple of passive observers
+	ob1 := mustCreateClientFor(t, e.Config().LCUrls[0].String())
+	defer ob1.Close()
+	ob2 := mustCreateClientFor(t, e.Config().LCUrls[0].String())
+	defer ob2.Close()
+
+	ctx := context.Background()
+	facts := make([]*ElectionFactory, 0)
```
Comment (daviddrysdale) [inline util/etcd/election_test.go]: Could just have
```go
facts := []*ElectionFactory{
    NewElectionFactory("serv1", client, "trees/"),
    NewElectionFactory("serv2", client2, "trees/"),
    NewElectionFactory("ob1", ob1, "trees/"),
    NewElectionFactory("ob2", ob2, "trees/"),
}
```

File: util/etcd/election_test.go [hunk 5]
```diff
@@ -54,3 +60,89 @@ func TestMasterElectionThroughCommonClient(t *testing.T) {
 		t.Fatalf("Close(20): %v", err)
 	}
 }
+
+func everyoneAgreesOnMaster(ctx context.Context, t *testing.T, want string, elections []util.MasterElection) {
+	t.Helper()
+	for _, e := range elections {
+		for {
+			log.Print("Getting master")
+			m, err := e.GetCurrentMaster(ctx)
```
Comment (daviddrysdale) [inline util/etcd/election_test.go]: Maybe `got, err` rather than `m, err` ?

File: util/etcd/election_test.go [hunk 6]
```diff
@@ -54,3 +60,89 @@ func TestMasterElectionThroughCommonClient(t *testing.T) {
...
+	wg := &sync.WaitGroup{}
+	for _, e := range elections[0:2] {
+		wg.Add(1)
+		go func(e util.MasterElection) {
+			if err := e.WaitForMastership(ctx); err != nil {
+				t.Errorf("WaitForMastership(10): %v", err)
+			}
+			id := e.(*MasterElection).instanceID
+
+			log.Printf("%v is master", id)
```
Comment (daviddrysdale) [inline util/etcd/election_test.go]: `t.Logf` is probably better here.

File: util/etcd/election_test.go [hunk 7]
```diff
@@ -54,3 +60,89 @@ func TestMasterElectionThroughCommonClient(t *testing.T) {
...
+			m, err := e.GetCurrentMaster(ctx)
+			if err == concurrency.ErrElectionNoLeader {
+				t.Error("No leader...")
+				time.Sleep(time.Second)
+				continue
+			} else if err != nil {
+				t.Fatalf("Failed to GetCurrentMaster: %v", err)
+			}
+			if m != want {
+				log.Printf("got %v want %v", m, want)
```
Comment (daviddrysdale) [inline util/etcd/election_test.go]: `t.Logf` preferred, but maybe just drop this log altogether?

File: server/log_operation_manager_test.go (hunk 1 above)
Comment (AlCutter) [reply server/log_operation_manager_test.go]: :)

File: util/election.go (hunk 2 above)
Comment (AlCutter) [reply util/election.go]: done x2

File: util/etcd/election_test.go (hunk 5 above)
Comment (AlCutter) [reply util/etcd/election_test.go]: done

File: util/etcd/election_test.go (hunk 7 above)
Comment (AlCutter) [reply util/etcd/election_test.go]: done

File: util/etcd/election_test.go (hunk 4 above)
Comment (AlCutter) [reply util/etcd/election_test.go]: Brain fart, thanks!

File: util/etcd/election_test.go (hunk 3 above)
Comment (AlCutter) [reply util/etcd/election_test.go]: done

File: util/etcd/election_test.go (hunk 6 above)
Comment (AlCutter) [reply util/etcd/election_test.go]: removed

File: util/etcd/election_test.go [hunk 8]
```diff
@@ -65,18 +64,16 @@ func everyoneAgreesOnMaster(ctx context.Context, t *testing.T, want string, elec
 	t.Helper()
 	for _, e := range elections {
 		for {
-			log.Print("Getting master")
-			m, err := e.GetCurrentMaster(ctx)
+			got, err := e.GetCurrentMaster(ctx)
 			if err == concurrency.ErrElectionNoLeader {
```
Comment (daviddrysdale) [inline util/etcd/election_test.go]: Does this test need updating for `util.ErrNoLeader`

Comment (AlCutter) [reply util/etcd/election_test.go]: thanks!

File: util/election.go [hunk 9]
```diff
@@ -67,6 +78,11 @@ func (ne *NoopElection) Close(ctx context.Context) error {
 	return nil
 }
 
+// GetCurrentMaster returns the string "It's you!"
+func (ne *NoopElection) GetCurrentMaster(ctx context.Context) (string, error) {
+	return "It's you!", nil
```
Comment (pavelkalinnikov) [inline util/election.go]: Should this return `ne.instanceID` instead?

Comment (AlCutter) [reply util/election.go]: Sure, done :)

**Comment by AlCutter** [comment]
PTAL

**Pull request merged**